framework/src/modules/interoperability/mainchain/module.ts

```diff
@@ -227,4 +244,191 @@ export class MainchainInteroperabilityModule extends BaseInteroperabilityModule
+			throw new Error(`chainData.name only uses the character set ${validNameCharset}.`);
+		}
+
+		// chainData.status is in set {CHAIN_STATUS_REGISTERED, CHAIN_STATUS_ACTIVE, CHAIN_STATUS_TERMINATED}.
+		if (!validStatuses.includes(chainData.status)) {
+			throw new Error(`chainData.status must be one of ${validStatuses.join(', ')}`);
+		}
+	}
+
+	private _verifyChannelData(ctx: GenesisBlockExecuteContext, chainInfo: ChainInfo) {
+		const mainchainTokenID = getMainchainTokenID(ctx.chainID);
+
+		const { channelData } = chainInfo;
+
+		// channelData.messageFeeTokenID == Token.getTokenIDLSK();
+		if (!channelData.messageFeeTokenID.equals(mainchainTokenID)) {
+			throw new Error(`channelData.messageFeeTokenID is not equal to Token.getTokenIDLSK().`);
+		}
+
+		// channelData.minReturnFeePerByte == MIN_RETURN_FEE_PER_BYTE_LSK.
+		if (channelData.minReturnFeePerByte !== MIN_RETURN_FEE_PER_BYTE_BEDDOWS) {
+			throw new Error(
+				`channelData.minReturnFeePerByte is not equal to MIN_RETURN_FEE_PER_BYTE_BEDDOWS.`,
+			);
+		}
+	}
+
+	private _verifyChainValidators(chainInfo: ChainInfo) {
+		const { chainValidators, chainData } = chainInfo;
+		const { activeValidators, certificateThreshold } = chainValidators;
+
+		// activeValidators must have at least 1 element and at most MAX_NUM_VALIDATORS elements
+		if (activeValidators.length === 0 || activeValidators.length > MAX_NUM_VALIDATORS) {
+			throw new Error(
+				`activeValidators must have at least 1 element and at most MAX_NUM_VALIDATORS elements.`,
+			);
+		}
+
+		// activeValidators must be ordered lexicographically by blsKey property
+		const sortedByBlsKeys = [...activeValidators].sort((a, b) => a.blsKey.compare(b.blsKey));
+		for (let i = 0; i < activeValidators.length; i += 1) {
+			if (!activeValidators[i].blsKey.equals(sortedByBlsKeys[i].blsKey)) {
+				throw new Error('activeValidators must be ordered lexicographically by blsKey property.');
+			}
+		}
+
+		// all blsKey properties must be pairwise distinct
+		const blsKeys = activeValidators.map(v => v.blsKey);
+		if (!bufferArrayUniqueItems(blsKeys)) {
+			throw new Error(`All blsKey properties must be pairwise distinct.`);
+		}
+
+		// for each validator in activeValidators, validator.bftWeight > 0 must hold
+		if (activeValidators.filter(v => v.bftWeight <= 0).length > 0) {
+			throw new Error(`validator.bftWeight must be > 0.`);
+		}
+
+		// let totalWeight be the sum of the bftWeight property of every element in activeValidators.
+		// Then totalWeight has to be less than or equal to MAX_UINT64
+		const totalWeight = activeValidators.reduce(
+			(accumulator, v) => accumulator + v.bftWeight,
+			BigInt(0),
+		);
+		if (totalWeight > MAX_UINT64) {
+			throw new Error(`totalWeight has to be less than or equal to MAX_UINT64.`);
+		}
+
+		// check that totalWeight//3 + 1 <= certificateThreshold <= totalWeight, where // indicates integer division
+		if (
+			totalWeight / BigInt(3) + BigInt(1) > certificateThreshold ||
```

Comment:
Just cross-checking: Is this integer division?

Reply:
Yes, BigInt division will alway return integer (rounding-down)